File: sdk/python-sdk/src/browser_agent/rpc.py
from __future__ import annotations

from dataclasses import replace

from .errors import BrowserAgentError, redact
from .events import RunStartedEvent, TaskResultEvent, UserTakeoverEvent
from .models import (
    BrowserAgentTaskResult,
    BrowserAgentTaskRunResult,
    BrowserAgentValidatorResult,
)

def record(value):
    if not isinstance(value, dict):
        raise BrowserAgentError(
            "PROTOCOL_ERROR", "CLI emitted invalid task-result parameters."
        )
    return value
def normalize_result(value) -> BrowserAgentTaskResult:
    source = record(value)
    task_id, status, raw_runs = (
        source.get("task_id"), source.get("status"), source.get("runs")
    )
    if not isinstance(task_id, str) or status not in ("completed", "failed") \
            or not isinstance(raw_runs, list):
        raise BrowserAgentError(
            "PROTOCOL_ERROR", "CLI emitted invalid task-result parameters."
        )
    runs = []
    for value in raw_runs:
        run = record(value)
        validator = record(run.get("validator"))
        if not isinstance(run.get("run_index"), int) \
                or not isinstance(run.get("completed"), bool) \
                or not isinstance(validator.get("ran"), bool) \
                or not isinstance(validator.get("success"), bool) \
                or not isinstance(validator.get("summary"), str):
            raise BrowserAgentError(
                "PROTOCOL_ERROR", "CLI emitted an invalid task run."
            )
        runs.append(BrowserAgentTaskRunResult(
            run["run_index"], run["completed"], run.get("data"),
            BrowserAgentValidatorResult(
                validator["ran"], validator["success"], validator["summary"]
            ),
        ))
    errors = source.get("errors")
    return BrowserAgentTaskResult(
        task_id, status, tuple(runs),
        tuple(item for item in errors if isinstance(item, str))
        if isinstance(errors, list) else (),
    )
def order(task_id: str) -> int:
    prefix, separator, suffix = task_id.partition("-")
    return int(suffix) if prefix == "task" and separator and suffix.isdigit() else 2**31
class RpcState:
    def __init__(self, run_id: str, secrets: list[str], paths: list[str]) -> None:
        self.run_id, self.secrets, self.paths = run_id, secrets, paths
        self._results: dict[str, BrowserAgentTaskResult] = {}
    @property
    def results(self) -> tuple[BrowserAgentTaskResult, ...]:
        return tuple(sorted(self._results.values(), key=lambda item: order(item.task_id)))
    def handle(self, message: dict[str, object]):
        if message.get("id") == 1:
            return self._accept(message)
        method = message.get("method")
        params = message.get("params")
        source = params if isinstance(params, dict) else {}
        if method == "browser-agent/status":
            category = source.get("category")
            if category not in ("authentication", "otp", "verification", "payment"):
                category = "other"
            return UserTakeoverEvent(
                self.run_id, str(source.get("task_id", "")),
                str(source.get("reason", "")), category
            )
        if method == "browser-agent/task_result":
            result = normalize_result(params)
            result = replace(result, errors=tuple(
                redact(error, self.secrets, self.paths) for error in result.errors
            ))
            self._results[result.task_id] = result
            return TaskResultEvent(self.run_id, result)
        if method == "browser-agent/all_tasks_completed":
            return "complete"
        if method == "browser-agent/error":
            raise BrowserAgentError(
                "PROCESS_EXITED",
                redact(str(source.get("message", "browser-agent failed.")),
                       self.secrets, self.paths),
            )
    def _accept(self, message: dict[str, object]):
        error = message.get("error")
        if isinstance(error, dict):
            data = error.get("data")
            code = data.get("code") if isinstance(data, dict) else None
            code = code if code in ("CONFIG_INVALID", "CHROME_NOT_FOUND") else "PROTOCOL_ERROR"
            raise BrowserAgentError(
                code,
                redact(str(error.get("message", "CLI rejected the run.")),
                       self.secrets, self.paths),
            )
        result = message.get("result")
        if not isinstance(result, dict) or result.get("accepted") is not True:
            raise BrowserAgentError("PROTOCOL_ERROR", "CLI did not accept the run.")
        return RunStartedEvent(self.run_id)

File: sdk/typescript-sdk/src/rpc.ts
import {
	BrowserAgentError,
	redact,
	type BrowserAgentErrorCode,
} from "./errors.js";
import type { RpcMessage } from "./protocol.js";
import type * as API from "./types.js";

const object = (value: unknown): Record<string, unknown> => {
	if (!value || typeof value !== "object" || Array.isArray(value))
		throw new BrowserAgentError(
			"PROTOCOL_ERROR",
			"CLI emitted invalid task-result parameters.",
		);
	return value as Record<string, unknown>;
};
function normalizeResult(value: unknown): API.BrowserAgentTaskResult {
	const input = object(value);
	if (
		typeof input.task_id !== "string" ||
		!["completed", "failed"].includes(String(input.status)) ||
		!Array.isArray(input.runs)
	)
		throw new BrowserAgentError(
			"PROTOCOL_ERROR",
			"CLI emitted invalid task-result parameters.",
		);
	const runs: API.BrowserAgentTaskRunResult[] = input.runs.map((value) => {
		const run = object(value);
		const validator = object(run.validator);
		if (
			typeof run.run_index !== "number" ||
			typeof run.completed !== "boolean" ||
			typeof validator.ran !== "boolean" ||
			typeof validator.success !== "boolean" ||
			typeof validator.summary !== "string"
		)
			throw new BrowserAgentError(
				"PROTOCOL_ERROR",
				"CLI emitted an invalid task run.",
			);
		return {
			runIndex: run.run_index,
			completed: run.completed,
			data: run.data ?? null,
			validator: {
				ran: validator.ran,
				success: validator.success,
				summary: validator.summary,
			},
		};
	});
	return {
		taskId: input.task_id,
		status: input.status as "completed" | "failed",
		runs,
		errors: Array.isArray(input.errors)
			? input.errors.filter(
					(item): item is string => typeof item === "string",
				)
			: [],
	};
}
const order = (taskId: string) => {
	const match = /^task-(\d+)$/.exec(taskId);
	return match ? Number(match[1]) : Number.MAX_SAFE_INTEGER;
};
const category = (value: unknown): API.UserTakeoverCategory =>
	["authentication", "otp", "verification", "payment"].includes(String(value))
		? (value as API.UserTakeoverCategory)
		: "other";

export class RpcState {
	readonly #results = new Map<string, API.BrowserAgentTaskResult>();
	constructor(
		private readonly runId: string,
		private readonly secrets: string[],
		private readonly paths: string[],
	) {}
	get results(): API.BrowserAgentTaskResult[] {
		return [...this.#results.values()].sort(
			(left, right) => order(left.taskId) - order(right.taskId),
		);
	}
	handle(
		message: RpcMessage,
	): API.BrowserAgentEvent | "complete" | undefined {
		if (message.id === 1) return this.#accept(message);
		const params = objectOrEmpty(message.params);
		if (message.method === "browser-agent/status")
			return {
				type: "user_takeover",
				runId: this.runId,
				taskId: String(params.task_id ?? ""),
				reason: String(params.reason ?? ""),
				category: category(params.category),
			};
		if (message.method === "browser-agent/task_result") {
			const result = normalizeResult(message.params);
			result.errors = result.errors.map((error) =>
				redact(error, this.secrets, this.paths),
			);
			this.#results.set(result.taskId, result);
			return { type: "task_result", runId: this.runId, result };
		}
		if (message.method === "browser-agent/all_tasks_completed") return "complete";
		if (message.method === "browser-agent/error")
			throw new BrowserAgentError(
				"PROCESS_EXITED",
				redact(
					String(params.message ?? "browser-agent failed."),
					this.secrets,
					this.paths,
				),
			);
	}
	#accept(message: RpcMessage): API.BrowserAgentEvent {
		if (message.error) {
			const code = message.error.data?.code;
			const allowed: BrowserAgentErrorCode[] = [
				"CONFIG_INVALID",
				"CHROME_NOT_FOUND",
			];
			throw new BrowserAgentError(
				allowed.includes(code as BrowserAgentErrorCode)
					? (code as BrowserAgentErrorCode)
					: "PROTOCOL_ERROR",
				redact(
					String(message.error.message ?? "CLI rejected the run."),
					this.secrets,
					this.paths,
				),
			);
		}
		if ((message.result as { accepted?: unknown })?.accepted !== true)
			throw new BrowserAgentError(
				"PROTOCOL_ERROR",
				"CLI did not accept the run.",
			);
		return { type: "run_started", runId: this.runId };
	}
}
const objectOrEmpty = (value: unknown): Record<string, unknown> =>
	value && typeof value === "object" && !Array.isArray(value)
		? (value as Record<string, unknown>)
		: {};
